# RNA sequencing preprocessing
> Preprocessing of RNA-sequencing including the quantification of transcriptome.

* [Package version](#package-version)
* [Transcriptome quantification](#transcriptome-quantification)
    + [Result](#result)
* [Reference](#reference)

## Package version
`salmon v1.8.0`

## Transcriptome quantification

### Command

The study used salmon as transcriptome quantifier. 

```sh
# salmon quantification
salmon quant -i $index -l A -1 $R1 -2 $R2 -p 8 --validateMappings --gcBias --seqBias --recoverOrphans -o $output
```

### Result

The mapping of results using Salmon ranges between 50% and 60% (Plot 1A). We noticed that many reads were removed due to quality issues or were not recorded within the index. In section 0, we suggested a low likelihood of foreign contamination. Therefore, considering these results together, we believe that the contamination we are observing here is possibly DNA or ncRNA not recorded in the index. This is a possible outcome of rRNA depletion, the method used for RNA extraction, so there isn't strong evidence to worry about quality issues for the RNA sequencing alignment result.

<br />
<p align="center">
  <img width="700" src="./Figure/Plot1A.png">
</p>

_**Plot1A. Alignment rate and component of RNA-seq library.** The alignment rate of each sample varies between 50% to 60%. Paired Mapped reads and Orphan Recovered reads are considered as usable reads (The two green)._

<br />

According to the theory of RNA-seq, long transcripts should have more RNA-seq library reads after read fragmentation. However, there are also studies that suggest that long transcripts are generally expressed at a low level due to the conservation of energy. Therefore, we plotted Plot1C to examine the distribution of TPM versus transcript length. We found that the overall distribution appears highly similar.

<br />
<p align="center">
  <img width="700" src="./Figure/Plot1B.png">
</p>

_**Plot1B. TPM versus transcript length.** The distribution shows to be highly similar._

## Reference

All command employed in this section can be found in [this script](../Script/1.%20RNA%20preprocessing.sh)

All plots in this section were created using [this script](../Script/1.%20Plotting.R)
