# ATAC sequencing preprocessing
> Preprocessing of ATAC-seq library includes the alignment, filtering, deduplication and peak calling.

* [Package version](#package-version)
* [Library trimming](#library-trimming)
    + [Command](#command)
* [Alignment](#alignment)
    + [Command](#command)
* [Post-alignment filtering](#post-alignment-filtering)
    + [Command](#command)
* [Methylation calling](#methylation-calling)
    + [Command](#command)
    + [Result](#result)
* [Reference](#reference)

## Package version
`trim_galore v0.6.7`
`cutadapt v4.1`
`bsmap v2.9.0`
`samtools v1.15.1`
`picard v2.27.0`
`MethylDackel v0.6.1`

## Library trimming

Trimming were performed to remove adapter, added in nucleotides due toshort reads and low quality of base calling. The fragmentation were performed using Tn5 transposase so the result could be suspected to suffer the same version of contamination of ATAC-seq. 

### Command

```sh
# WGBS library trimming
trim_galore --paired --output_dir $trim_dir $R1 $R2
```

## Alignment

The alignment were performed using BSMAP which is one of the most ideal in choosing between alignment rate, speed and computational requirement. But it would be ideal to change to bismark as it allows the whole study to use bowtie2 based method.

### Command

```sh
# WGBS alignment
bsmap -a $R1 -b $R2 -d GRCh38.primary_assembly.genome.fa -x 1000 -o $Output.sam

# Converting to bam file to save storage
# Remove all uninterested regions as it signficantly increase the error chance in following procedure.
samtools view -bS $Output.sam chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY > $Output.bam
```

## Post-alignment filtering

### Command

```sh
# Mark duplicates
picard MarkDuplicates I=$Output.bam O=$Output_dedup.bam M=$Output_dedup.txt
```

## Methylation calling

As the fragmentation of WGBS library were performed using Tn5 transposase, there is a biased region of 9bp at the 5' end of all reads. To be precise, we went through MethylDackel mbias and decided to remove 10bp from the 5' end as the 10th bp seems to also suffer a certain level of bias.

### Command

```sh
# Using MethylDackel to extract methylation information
MethylDackel extract -@ $Thread --OB 10,0,10,0 --OT 0,140,0,140 GRCh38.primary_assembly.genome.fa $Output_dedup.bam
```

### Result




## Reference
