# Differential accessibility analysis
> Using DESeq2 to compare the change in gene expression due to PMA treatment. 

* [Package version](#package-version)
* [Differential expression](#differential-expression)
    + [Command](#command)
    + [Result](#result)
* [Enrichment analysis](#enrichment-analysis)
    + [Command](#command)
    + [Result](#result)
* [Reference](#reference)

## Package version
`DESeq2 v1.34.0`
`tximport v1.22.0`
`biomaRt v2.50.3`
`tidyverse v1.3.1`


## Differential expression

### Command
Importing all transcriptome quantification from salmon into a gene-level quanitification with a dataframe format as input for DEseq2.

```R
library("tximport")

# Listing the path to each file.
files <- c('../Input/0hr_rep1/quant.sf',
           '../Input/0hr_rep2/quant.sf',
           '../Input/24hr_rep1/quant.sf',
           '../Input/24hr_rep2/quant.sf')
          
# Covert tx into gene          
txi.salmon <- tximport(files = files,
                       type = "salmon",
                       txOut = FALSE,
                       tx2gene = tx2gene,
                       ignoreAfterBar = TRUE)
                      
# Providing condition name to each replicates
sampleTable <- data.frame(condition = factor(rep(c("control", "treated"),each = 2)))
rownames(sampleTable) <- colnames(txi.salmon$counts)  
```

Running DESeq2 for differential expression detection.

```R
library("DESeq2")
library("IHW")

# Importing the result from tximport to DESeq2
dds <- DESeqDataSetFromTximport(countData = txi.salmon,
                                colData = sampleTable,
                                design = ~ condition) 

# Perform Differential expression analysis using Wald test
dds <- DESeq(object = dds,
             test = 'Wald')

# Extracting results from the analysis, specifiy the significant cut is 0.05
result <- results(object = dds,
                  filterFun = ihw,
                  alpha = 0.05)

# Performing l2FC shrinkage to remove bias.
result <- lfcShrink(dds = dds,
                    coef = "condition_treated_vs_control",
                    type="apeglm",
                    res = result)
```

At this point, we will be done with DESeq2. Now it to make it easier for intepretation

```R
library(GenomicFeatures)
library(biomaRt)

# Only keep the stable gene ID
result$Row.names <- gsub("(ENSG[0-9]+)\\.[0-9]+", "\\1",result$Row.names) 
# Getting the gene number of name
G_list = getBM(filters = "ensembl_gene_id",
               attributes = c("ensembl_gene_id","hgnc_symbol",'entrezgene_id'),
               values = result$Row.names,
               mart= biomart)
# Adding the name and DESeq2 result into one dataframe
result = merge(result,G_list,by.x="Row.names",by.y="ensembl_gene_id") 
# Putting the accession number into transcripts without a official name
result$hgnc_symbol = ifelse(result$hgnc_symbol =='',result$Row.names,result$hgnc_symbol)  
# Remove duplicate recordings
result = result %>% distinct(Row.names,hgnc_symbol,log2FoldChange,.keep_all = T)

# Certain genes were not recprded in the biomart, so we use two databases
result$id = mapIds(org.Hs.eg.db,keys = result$Row.names,column = 'ENTREZID',keytype = 'ENSEMBL',multiVals = 'first')
# Combine the two results into one column
for(i in 1:nrow(result)) {
  if(is.na(result[i,'entrezgene_id'])){
    result[i,'entrezgene_id'] = result[i,'id']
  } 
}
```

Now adding in cutoff values to determine the final results

```R
library(tidyverse)
DEG = result %>% dplyr::filter(padj <0.05 & baseMean > 20 & abs(log2FoldChange) > 2) 

result$con = "non-significant"
result$con[result$log2FoldChange > 2 & result$padj < 0.05] = "upregulated"
result$con[result$log2FoldChange < -2 & result$padj < 0.05] = "downregulated"
```

### Result

To shows the number of differential expressed genes and the distribution, we will be plotting a volcanoi lot for DEGs to show there parameters. 

```R
library(tidyverse)
library(gridExtra)

# Making all 0 padj to the lowest possible DESeq2 padj due to DESeq2 limitation
resdata_24hr$padj = ifelse(resdata_24hr$padj == 0,1e-303,resdata_24hr$padj)

# Plotting
ggplot(x,aes(x=log2FoldChange,y=-log10(padj),color = con)) + geom_point() +
    theme_bw() +                                                                  
    theme(text = element_text(size = 20),legend.position = "none",rect = element_rect(fill = "transparent")) +
    scale_colour_manual(values = c("#4995C6","#C0C0C0","#B9181A")) +                       # Change the colour of dots
    labs(x = expression("Change in Gene Expression (Log"[2]*" Fold Change)"),y = expression("-log"[10]*"(p adjusted)"),color = "Gene status") + #naming
    scale_y_continuous(expand = c(0,0),limits = c(0,310))                                         # Remove the gap between the results and x margin

ggsave("VolcanoPlot.png",width = 18,height = 10,dpi = 1200)
```

<br />
<p align="center">
  <img width="1010" height="558" src="https://github.com/Yifan-bio/msc/blob/1fb309d9561fa7637da54930f01d4f31f3de1667/Master/Image/VolcanoPlot.png">
</p>

_**Figure 3.2. Volcano Plot.** L_


## Enrichment analysis

Ads now we have a bunch of DEGs, we will like to obtain the biological meaning of these DEGs through running enrichment analysis. But kegg pathways shows a limited of understanding on the impact so we will only be focusing on Gene ontology of the genes.

### Command

### Result

## Reference

