# Quality control
> The file records all quality control performed during the selection of datasets.

* [Package version](#package-version)
* [Employed dataset](#employed-dataset)
* [Queried dataset](#queried-dataset)
* [Reference](#reference)

## Package version
`fastqc v0.11.9`
`multiqc v1.14`


## Employed dataset

The following figure shows the quality of all fastq file that were incoporated during the study. This study includes RNA-seq, ATAC-seq and WGBS for 0 hour (untreat) and 24 hour 100ng/ml (~162nM) PMA treated THP-1 cells. The dataset are recorded under PRJNA533829 / GSE130096.

```R
library(stringr)
library(data.table)

# Define a function to extract the per base quality data from a single FastQC report
extract_per_base_quality <- function(fastqc_data) {
  # Find the start and end indices of the "Per base sequence quality" section
  start <- str_locate(fastqc_data, ">>Per base sequence quality")[1]
  end <- str_locate(fastqc_data, ">>END_MODULE")[1] - 1

  # Extract the lines for the section
  section_data <- str_split(fastqc_data[start:end], "\n")[[1]]

  # Split the lines into columns by whitespace
  column_names <- c("Base", str_split(section_data[2], "\\s+")[[1]])
  data <- lapply(section_data[-c(1:2)], function(line) str_split(line, "\\s+")[[1]])
  data <- data[data$sapply(length) == length(column_names), ]

  # Create a data table with the section data
  dt <- data.table(data, keep.rownames = FALSE)
  setnames(dt, old = names(dt), new = column_names)
  dt[, Base := as.numeric(Base)]

  # Return the data table
  return(dt)
}

# Define a function to extract the per base quality data from all FastQC reports in a directory
extract_per_base_quality_from_dir <- function(dir_path) {
  dfs <- list()
  for (filename in dir(dir_path, pattern = ".zip$")) {
    fastqc_path <- file.path(dir_path, sub("\\.zip$", "_fastqc/fastqc_data.txt", filename))
    if (file.exists(fastqc_path)) {
      fastqc_data <- readLines(fastqc_path, warn = FALSE)
      df <- extract_per_base_quality(fastqc_data)
      df$Sample <- gsub("\\.zip$", "", filename)
      dfs <- c(dfs, list(df))
    }
  }
  if (length(dfs) == 0) {
    return(NULL)
  } else {
    return(rbindlist(dfs))
  }
}

# Example usage
per_base_quality_df <- extract_per_base_quality_from_dir("/path/to/fastqc_outputs")

```

<br />
<p align="center">
  <img width="460" height="300" src="https://github.com/Yifan-bio/msc/blob/8580c17de235bce9403d6e215b2cba4cb634416f/Master/Image/Figure%200.1.A.png">
  <img width="460" height="300" src="https://github.com/Yifan-bio/msc/blob/8580c17de235bce9403d6e215b2cba4cb634416f/Master/Image/Figure%200.1.B.png">
</p>

_**Figure 3.1. All employed datasets present qualified sequencing quality.** The graph separately represents a) the per base sequencing quality and b) the mean sequence quality of all reads. All sequencing libraries were derived from 100 ng/ml PMA treated and untreated THP-1 cells collected from PRJNA53382. Expected contamination were found for both ATAC-seq library and WGBS library._

<br />

WGBS and ATAC-seq has shown to suffer to the technique specific quality challenges which was accounted for during the processing of dataset. This bias includes high adapter sequencing rate and high duplication rate introudced by Tn5 tranposase during library extraction. 

## Queried dataset

Other datasets such as PRJNA379744 were also evaluated but shows to suffer with a none human source DNA contamination.

## Reference

