# Quality control
> The file records all quality control performed during the selection of datasets.

* [Package version](#package-version)
* [Employed dataset](#employed-dataset)
  + [Results](#results)
* [Queried dataset](#queried-dataset)
* [Reference](#reference)

## Package version
`fastqc v0.11.9`
`multiqc v1.14`


## Employed dataset

The following figure shows the quality of all fastq file that were incoporated during the study. This study includes RNA-seq, ATAC-seq and WGBS for 0 hour (untreat) and 24 hour 100ng/ml (~162nM) PMA treated THP-1 cells. The dataset are recorded under PRJNA533829 / GSE130096.

| Accession Number | Technique | Condition | Number of Reads (Million) |
| ------------------ | ------------------: | ------------------: | ----------------: |
| SRR8932925 | ATAC | 0 hour | 134.2 |
| SRR8932926 | ATAC | 24 hour | 128.2 |
| SRR8932927 | ATAC | 0 hour | 101.5 |
| SRR8932928 | ATAC | 24 hour | 101.6 |
| SRR8932929 | RNA | 24 hour | 70.7 |
| SRR8932930 | RNA | 0 hour | 77.2 |
| SRR8932934 | RNA | 24 hour | 47.4 |
| SRR8932935 | RNA | 0 hour | 82.1 |
| SRR8932923 | WGBS | 0 hour | 446.6 |
| SRR8932924 | WGBS | 24 hour | 458.9 |


### Results

Overall results for the sequencing quality were found to be sufficient (Plot0A & Plot0B). WGBS shows to suffer a level of sequencing quality which is commonly observed in bisulfite sequencing reads. Additionally, RNA sequencing results were found to be included a contamination of reads which 
RNA-seq results were shown contain a certain level of contamination as several peaks were found for the fastq GC content (Plot0C). WGBS results shows to suffer with the end quality which is commonly observed for Bisulfite sequencing results. 

WGBS and ATAC-seq has shown to suffer to the technique specific quality challenges which was accounted for during the processing of dataset. This bias includes high adapter sequencing rate and high duplication rate introudced by Tn5 tranposase during library extraction. 

<br />
<p align="center">
  <img width="700" src="https://user-images.githubusercontent.com/87715781/226893072-505b1442-db78-4a80-a29f-c83b3d4eb53c.png">
</p>

_**Plot0A. All employed datasets present qualified per base quality.** The dataset were collected under bioproject PRJNA53382. The colours separately represent RNA-seq datasets (orange), ATAC-seq datasets (blue) and WGBS (green)._

<br />
<p align="center">
  <img width="700" src="https://user-images.githubusercontent.com/87715781/226893884-11d3881d-fbf9-4c66-a0f0-f44f2ffa46b9.png">
</p>

_**Plot0B. All employed datasets present qualified per sequence quality.** The dataset were collected under bioproject PRJNA53382. The colours separately represent RNA-seq datasets (orange), ATAC-seq datasets (blue) and WGBS (green)._

<br />
<p align="center">
  <img width="700" src="https://user-images.githubusercontent.com/87715781/226893914-f34c1205-edaa-4abd-adaa-91d53c9630dd.png">
</p>

_**Plot0C. GC content of all datasets shows a potential contamination of RNA-seq.** The dataset were collected under bioproject PRJNA53382. The colours separately represent RNA-seq datasets (orange), ATAC-seq datasets (blue) and WGBS (green)._

## Queried dataset

Other datasets such as PRJNA379744 were also evaluated but shows to suffer with a none human source DNA contamination.

## Reference

