# Quality control
> The file records all quality control performed during the selection of datasets.

* [Package version](#package-version)
* [Employed dataset](#employed-dataset)
  + [Results](#results)
* [Queried dataset](#queried-dataset)
* [Reference](#reference)

## Package version
`fastqc v0.11.9`
`multiqc v1.14`


## Employed dataset

The following figure shows the quality of all fastq file that were incoporated during the study. This study includes RNA-seq, ATAC-seq and WGBS for 0 hour (untreat) and 24 hour 100ng/ml (~162nM) PMA treated THP-1 cells. The dataset are recorded under PRJNA533829 / GSE130096.

| Accession Number | Technique | Condition | Number of Reads (Million) |
| ------------------ | ------------------: | ------------------: | ----------------: |
| SRR8932925 | ATAC | 0 hour | 134.2 |
| SRR8932926 | ATAC | 24 hour | 128.2 |
| SRR8932927 | ATAC | 0 hour | 101.5 |
| SRR8932928 | ATAC | 24 hour | 101.6 |
| SRR8932929 | RNA | 24 hour | 70.7 |
| SRR8932930 | RNA | 0 hour | 77.2 |
| SRR8932934 | RNA | 24 hour | 47.4 |
| SRR8932935 | RNA | 0 hour | 82.1 |
| SRR8932923 | WGBS | 0 hour | 446.6 |
| SRR8932924 | WGBS | 24 hour | 458.9 |


### Results

```R
library(tidyverse)

# Per base quality
temp =  read_tsv("./Input/fastqc_per_base_sequence_quality_plot.tsv")
# Renaming the files
colnames(temp) = c("Pos",paste0("ATAC",1:8),paste0("WGBS",1:4),paste0("RNA",1:8))
# Convert to long dataframe
temp = temp %>% gather(key = "variable",value = "value",-Pos)
# Allowing all same tech to be in one colour
temp$colour = gsub('[[:digit:]]+', '',temp$variable)
temp=na.omit(temp)
# Plot
plot1 = ggplot(temp,aes(x = Pos,y=value),) +
        geom_line(aes(group = variable,color = colour),show.legend = FALSE) +
        scale_x_continuous(limits = c(0,150),expand = c(0,0)) +
        scale_y_continuous(limits = c(0,41),expand = c(0,0)) +
        xlab("Position in read(bp)") +
        ylab("Phred Score") +
        theme_bw() +
        scale_color_manual(values = c("#1f88b4","#ff7f0e","#2ca02c"))

# Per sequence quality
temp = read_tsv("./Input/fastqc_per_sequence_quality_scores_plot.tsv")
# Renaming the files
colnames(temp) = c("Score",paste0("ATAC",1:8),paste0("WGBS",1:4),paste0("RNA",1:8))
# Convert to long dataframe
temp = temp %>% gather(key = "variable",value = "value",-Score)
temp$colour = gsub('[[:digit:]]+', '',temp$variable)

plot2 = ggplot(temp,aes(x = Score,y=value)) +
        geom_line(aes(group = variable,color = colour),show.legend = FALSE) +
        scale_x_continuous(limits = c(0,40),expand = c(0,0)) +
        scale_y_continuous(limits = c(0,150000000),expand = c(0,0)) +
        xlab("Phred Score") +
        ylab("Number of reads") +
        theme_bw() +
        scale_color_manual(values = c("#1f88b4","#ff7f0e","#2ca02c"))

ggpubr::ggarrange(plot1,plot2)
ggsave(dpi=900,"fastqc.png")
```

<br />
<p align="center">
  <img width="1180" height="349" src="https://github.com/Yifan-bio/msc/blob/a18157cb5666faf5d52caf6f94ca46d758a48647/Master/Image/0.1.%20fastqc.png">
</p>

_**Figure 3.1. All employed datasets present sufficient sequencing quality.** The graph separately represents a) the per base sequencing quality and b) the mean sequence quality of the dataset collected from PRJNA53382. The colours separately represent RNA-seq datasets (orange), ATAC-seq datasets (blue) and WGBS (green). Although all datasets exhibit sufficient sequencing quality, certain biases due to library preparation can be observed.y._

<br />

WGBS and ATAC-seq has shown to suffer to the technique specific quality challenges which was accounted for during the processing of dataset. This bias includes high adapter sequencing rate and high duplication rate introudced by Tn5 tranposase during library extraction. 

## Queried dataset

Other datasets such as PRJNA379744 were also evaluated but shows to suffer with a none human source DNA contamination.

## Reference

