# ATAC sequencing preprocessing
> Preprocessing of ATAC-seq library includes the alignment, filtering, deduplication and peak calling.

* [Package version](#package-version)
* [Library trimming](#library-trimming)
    + [Command](#command)
    + [Result](#result)
* [Alignment](#alignment)
    + [Command](#command)
    + [Result](#result)
* [Post-alignment filtering](#post-alignment-filtering)
    + [Command](#command)
    + [Result](#result)
* [Methylation calling](#methylation-calling)
    + [Command](#command)
    + [Result](#result)
* [Reference](#reference)

## Package version
`trim_galore v0.6.7`
`cutadapt v4.1`
`bsmap v2.9.0`
`samtools v1.15.1`
`picard v2.27.0`
`MethylDackel v0.6.1`

## Library trimming

### Command

Trimming were performed to remove adapter, added in nucleotides due toshort reads and low quality of base calling. The fragmentation were performed using Tn5 transposase so the result could be suspected to suffer the same version of contamination of ATAC-seq. 

```sh

```

### Result


## Alignment

### Command

The alignment were performed using BSMAP which is one of the most ideal in choosing between alignment rate, speed and computational requirement. 

```sh
# ATAC-seq alignment
bowtie2 -a -b -d
```

### Result

## Post-alignment filtering

### Command

```sh

```

### Result

## Methylation calling

### Command

```sh

```

### Result




## Reference
