# ATAC sequencing preprocessing
> Preprocessing of ATAC-seq library includes the alignment, filtering, deduplication and peak calling.

* [Package version](#package-version)
* [Library trimming](#library-trimming)
    + [Command](#command)
    + [Plotting](#plotting)
* [Alignment](#alignment)
    + [Command](#command)
    + [Plotting](#plotting)
* [Post-alignment filtering](#post-alignment-filtering)
    + [Command](#command)
    + [Plotting](#plotting)
* [Methylation calling](#methylation-calling)
    + [Command](#command)
    + [Plotting](#plotting)
* [Reference](#reference)

## Package version
`trim_galore v0.6.7`
`cutadapt v4.1`
`bsmap v2.9.0`
`samtools v1.15.1`
`picard v2.27.0`

## Library trimming

### Command

The trimming were performed to remove adapter and random hexamers added during sequencing. Majority of the ATAC-seq reads are around 70bp as an accessible region between nucleosomes as around 75bp. So by trimming we can remove the low quality, random hexamers added in for short fragments and also adpater sequences.

```sh

```

### Plotting


## Alignment

### Command

```sh
# ATAC-seq alignment
bowtie2 --very-sensitive --end-to-end -p 8 --dovetail --no-mixed -X 2000 -t -x $Index -1 $Read1 -2 $Read2 > ${prefix}.bowtie2.log | samtools sort -@ 4 -O bam -o ${prefix}.bam
```

### Plotting

## Post-alignment filtering

### Command

```sh

```

### Plotting

## Methylation calling

### Command

```sh

```

### Plotting




## Reference
