# Differential accessibility analysis
> Using DESeq2 to compare the change in gene expression due to PMA treatment. 

* [Package version](#package-version)
* [Differential expression](#differential-expression)
    + [Command](#command)
    + [Result](#result)
* [Enrichment analysis](#enrichment-analysis)
    + [Command](#command)
    + [Result](#result)
* [Reference](#reference)

## Package version
`DESeq2 v1.34.0`
`tximport v1.22.0`
`biomaRt v2.50.3`
`tidyverse v1.3.1`


## Differential expression

### Command
Importing all transcriptome quantification from salmon into a gene-level quanitification with a dataframe format as input for DEseq2.

```R
library("tximport")

# Listing the path to each file.
files = c('../Input/0hr_rep1/quant.sf',
          '../Input/0hr_rep2/quant.sf',
          '../Input/24hr_rep1/quant.sf',
          '../Input/24hr_rep2/quant.sf')
          
# Covert tx into gene          
txi.salmon = tximport(files,
                      type = "salmon",
                      txOut = F,
                      tx2gene = tx2gene,
                      ignoreAfterBar = T)
                      
# Providing condition name to each replicates
sampleTable <- data.frame(condition = factor(rep(c("control", "treated"),each = 2)))
rownames(sampleTable) <- colnames(txi.salmon$counts)  
```

Running DESeq2 for differential expression detection.

```R
library("DESeq2")
library("IHW")

dds <- DESeqDataSetFromTximport(countData = txi.salmon,
                                colData = sampleTable,
                                design = ~ condition) 

dds_24hr <- DESeq(dds_24hr,test = 'Wald')                                       # Differential expression analysis


```

### Result

## Enrichment analysis

### Command

### Result

## Reference

