# Differential accessibility analysis
> Using DESeq2 to compare the change in gene expression due to PMA treatment. 

* [Package version](#package-version)
* [Differential expression](#differential-expression)
    + [Command](#command)
    + [Result](#result)
* [Enrichment analysis](#enrichment-analysis)
    + [Command](#command)
    + [Result](#result)
* [Reference](#reference)

## Package version
`DESeq2 v1.34.0`
`tximport v1.22.0`
`biomaRt v2.50.3`
`tidyverse v1.3.1`


## Differential expression

### Command
Importing all transcriptome quantification from salmon into a gene-level quanitification with a dataframe format as input for DEseq2.

```R
library("tximport")

# Listing the path to each file.
files <- c('../Input/0hr_rep1/quant.sf',
           '../Input/0hr_rep2/quant.sf',
           '../Input/24hr_rep1/quant.sf',
           '../Input/24hr_rep2/quant.sf')
          
# Covert tx into gene          
txi.salmon <- tximport(files = files,
                       type = "salmon",
                       txOut = FALSE,
                       tx2gene = tx2gene,
                       ignoreAfterBar = TRUE)
                      
# Providing condition name to each replicates
sampleTable <- data.frame(condition = factor(rep(c("control", "treated"),each = 2)))
rownames(sampleTable) <- colnames(txi.salmon$counts)  
```

Running DESeq2 for differential expression detection.

```R
library("DESeq2")
library("IHW")

# Importing the result from tximport to DESeq2
dds <- DESeqDataSetFromTximport(countData = txi.salmon,
                                colData = sampleTable,
                                design = ~ condition) 

# Perform Differential expression analysis using Wald test
dds <- DESeq(object = dds,
             test = 'Wald')

# Extracting results from the analysis, specifiy the significant cut is 0.05
result <- results(object = dds,
                  filterFun = ihw,
                  alpha = 0.05)

# Performing l2FC shrinkage to remove bias.
result <- lfcShrink(dds = dds,
                    coef = "condition_treated_vs_control",
                    type="apeglm",
                    res = result)
```

At this point, we will be done with DESeq2. Now it to make it easier for intepretation
```R
library(GenomicFeatures)
library(biomaRt)

# Only keep the stable gene ID
result <- gsub("(ENSG[0-9]+)\\.[0-9]+", "\\1",result$Row.names) 
# Getting the gene number of name
G_list = getBM(filters = "ensembl_gene_id",
               attributes = c("ensembl_gene_id","hgnc_symbol",'entrezgene_id'),
               values = result$Row.names,
               mart= biomart)


```


### Result

To shows the number of differential expressed genes and the distribution, we will be plotting a volcanoi lot for DEGs to show there parameters. 

```R

```

## Enrichment analysis

Ads now we have a bunch of DEGs, we will like to obtain the biological meaning of these DEGs through running enrichment analysis. But kegg pathways shows a limited of understanding on the impact so we will only be focusing on Gene ontology of the genes.

### Command

### Result

## Reference

