---
title: "Multi-omics analysis reveals early genomic responder to monocyte differentiation inducer"
output: html_notebook
date: "2024-03-16"
---

> After each block only the required objects will be stored as rds object, can clear the image after every block

# Section One

## Running the RNA analysis for PRJNA533

> This section is written to allow the running of differential analysis of RNA sequencing for THP-1 cells that were treated with 100ng/ml of PMA for 24 hours.

### Running the preparation of RNA seuqencing analysis

* Input: GTF file
* Using the GTF file to create a table to convert gene names, transcript ID and gene ID
* Output: tx2gene.rds; gtf_file.rds

```{r}
# Running preparation of RNA sequencing analysis 

library(GenomicFeatures)

# This section allows the creation of tx2gene table that is used for ID conversion for DESeq2
tx2gene_creation = function(gtf_file) {
  # Intaking the gtf file as binary object for R
  temp1 = GenomicFeatures::makeTxDbFromGFF(file = gtf_file)
  # Extracting the required information
  temp2 = AnnotationDbi::keys(x = temp1,
                              keytype = 'TXNAME')
  # Selecting the required information
  temp = AnnotationDbi::select(temp1,
                               keys = temp2,
                               columns = 'GENEID',
                               keytype = 'TXNAME')
  return(temp)
}

# Running the function to create the refence table.
tx2gene = tx2gene_creation('../../Supporting_file/gencode.v43.primary_assembly.annotation.gtf')
saveRDS(tx2gene,"../rds/tx2gene.rds")

# Creating a ID to symbol conversion table using raw GTF file
gtf_file <- rtracklayer::import('../../Supporting_file/gencode.v43.primary_assembly.annotation.gtf')
gtf_file <- as.data.frame(gtf_file)
gtf_file <- unique(gtf_file[,c("gene_id","gene_name")])
saveRDS(gtf_file,"../rds/gtf_file.rds")

```

### Running the differential expression analysis of the THP-1 cells

* Input: tx2gene.rds; salmon quant file; gtf_file.rds
* Running the differential analysis using salmon results
* Output: resdata.rds; expressed.rds; DEG.rds

```{r}
# Script to run the differential analysis for RNA-seq

library("DESeq2")
library("tximport")
library("tidyverse")

#### Function One: Running DESeq2  ####

dds_creation = function(files,tx2gene,input_condition){
  
  # Import files into tximport
  txi.salmon <- tximport(files = files,
                         type = "salmon",
                         txOut = FALSE,
                         tx2gene = tx2gene,
                         ignoreAfterBar = TRUE
                         )
  
  # Creating a table with the metainfo
  sampleTable <- data.frame(condition = input_condition)

  # Combining the metainfo with the tximport results
  rownames(sampleTable) <- colnames(txi.salmon$counts) 

  # Importing information into DESeq2 format
  dds <- DESeqDataSetFromTximport(txi = txi.salmon, 
                                  colData = sampleTable,
                                  design = ~ condition)

  # Running the differential expression analysis 
  dds <- DESeq(object = dds,
               test = 'Wald')
  
  return(dds)
  
}

#### Function Two: Extracting usable results and annotate ####

resdata_creation <- function(dds,gtf_file = FALSE,gtf_rds = FALSE){
  
  # Getting the results out of DEseq2
  result = lfcShrink(dds = dds,
                     coef = "condition_treated_vs_control",
                     type = "apeglm"
  )
  
  gtf_file <- readRDS(gtf_rds)
  
  # Putting all important information into a single object
  res <- merge(as.data.frame(result),
               as.data.frame(counts(dds,normalized=TRUE)),
               by = "row.names",
               sort = FALSE)
  
  # Putting the gene symbol with the gene id
  res <- merge(x = res,
               y = gtf_file,
               by.x = "Row.names",
               by.y = "gene_id")
  
  # Getting gene stable version number
  res$Row.names <- gsub("(ENSG[0-9]+)\\.[0-9]+", "\\1",res$Row.names) 
  
  return(res)
}

############### Running differential analysis ##################

dds = dds_creation(files = c('../Step 1 - Preprocessing/THP1_Salmon/SRR8932930_0hrPMA_rep1/quant.sf',
                             '../Step 1 - Preprocessing/THP1_Salmon/SRR8932935_0hrPMA_rep2/quant.sf',
                             '../Step 1 - Preprocessing/THP1_Salmon/SRR8932934_24hrPMA_rep1/quant.sf',
                             '../Step 1 - Preprocessing/THP1_Salmon/SRR8932929_24hrPMA_rep2/quant.sf'),
                   tx2gene = readRDS("../Step 3 - Differential Analysis/tx2gene.rds"),
                   input_condition = factor(rep(c("control", "treated"),each = 2)))

resdata = resdata_creation(dds = dds,
                           gtf_rds = "../Step 3 - Differential Analysis/gtf_file.rds")

expressed = resdata[resdata$baseMean > 20,]

DEG = expressed[expressed$padj < 0.05 & abs(expressed$log2FoldChange) > 2,]

saveRDS(resdata,"../rds/RNA_resdata.rds")
saveRDS(expressed,"../rds/RNA_expressed.rds")
saveRDS(DEG,"../rds/RNA_DEGs.rds")
```

### Running the over-representaion analysis of differentially expressed genes

* Input: DEG.rds; expressed.rds
* Outputï¼šORA.rds

```{r}

library("clusterProfiler")
library("org.Hs.eg.db")

expressed = readRDS("../Step 3 - Differential Analysis/RNA_expressed.rds")
DEG = readRDS("../Step 3 - Differential Analysis/RNA_DEGs.rds")

# Function for clusterProfiler
ORA_GO = function(background,sig_gene,term = 'ALL') {
  # Setting the universal genes
  BG = as.character(background$Row.names)
  # Creating the genelist for ORA
  genelist = sig_gene$log2FoldChange
  names(genelist) = sig_gene$Row.names
  genelist = sort(genelist,decreasing = TRUE)
  gene = names(genelist)[abs(genelist) > 2]
  # Enrichment analysis using clusterProfiler
  res <- enrichGO(gene = gene, 
                  universe = BG,
                  OrgDb = 'org.Hs.eg.db',
                  ont = term,
                  keyType = 'ENSEMBL',
                  pAdjustMethod = "BH",
                  pvalueCutoff = 0.01,
                  readable = TRUE)
  return(res)
}

ORA = ORA_GO(background = expressed,sig_gene = DEG,term = "ALL")
saveRDS(ORA,"../rds/RNA_ORA.rds")
```

## Running the ATAC-seq analysis for PRJNA

> This section is written to allow the running of differential analysis of ATAC sequencing for THP-1 cells that were treated with 100 ng/ml of PMA for 24 hours.

### Running the differential accessibility analysis of the THP-1 cells

* Input: Genrich peak; BAM file
* Output: regdata.rds; DAR.rds; DAR.bed

```{r}
# Script to run the differential analysis for ATAC-seq

library("DiffBind")
library(tidyverse)

# Putting in the metadata
sampleTbl = data.frame(sampleID = c('control1','control2','treated1','treated2'),
                       Cell = 'THP1',
                       Factor = 'PMA',
                       Condition= c('Control','Control','Treated','Treated'),
                       Replicate = c("1","2","1","2"),
                       bamReads = c('../Step 1 - Preprocessing/THP1_bowtie2/untreat_rep1.rmChrM.dedup.filter.bam',
                                    '../Step 1 - Preprocessing/THP1_bowtie2/untreat_rep2.rmChrM.dedup.filter.bam',
                                    '../Step 1 - Preprocessing/THP1_bowtie2/treat_rep1.rmChrM.dedup.filter.bam',
                                    '../Step 1 - Preprocessing/THP1_bowtie2/treat_rep2.rmChrM.dedup.filter.bam'),
                       Peaks = c('../Step 1 - Preprocessing/THP1_Genrich/untreat_rep1.narrowPeak',
                                 '../Step 1 - Preprocessing/THP1_Genrich/untreat_rep2.narrowPeak',
                                 '../Step 1 - Preprocessing/THP1_Genrich/treat_rep1.narrowPeak',
                                 '../Step 1 - Preprocessing/THP1_Genrich/treat_rep2.narrowPeak'),
                       ScoreCol = 5,
                       LowerBetter = FALSE
)

# Importing the data
ATAC <- dba(sampleSheet=sampleTbl)

# Perform count on each peak using the BAM file
ATAC_count <- dba.count(ATAC,minOverlap = 2,
                        score = DBA_SCORE_NORMALIZED,
                        bUseSummarizeOverlaps = TRUE
)

saveRDS(ATAC_count,"../rds/ATAC_count.rds")

################################ Running DiffBind ##############################
################################################################################

# Differential accessibility analysis
ATAC_contrast <- dba.contrast(ATAC_count, categories=DBA_CONDITION,minMembers = 2)

# Perform differential affinitiy analysis
ATAC_analyze <- dba.analyze(ATAC_contrast,method = DBA_ALL_METHODS)

# Extracting the results
ATAC_report <- dba.report(DBA = ATAC_analyze,
                          DataType = DBA_DATA_FRAME,
                          method = DBA_DESEQ2,
                          contrast = 1,
                          th = 1)

saveRDS(ATAC_report,"../rds/ATAC_regdata.rds")

# Extracting usable regions
DAR = ATAC_report %>% dplyr::filter(abs(Fold) > 2 & FDR < 0.01)
row.names(DAR) = 1:nrow(DAR)
saveRDS(DAR,"../rds/ATAC_DAR.rds")

# Output for GREAT analysis
write_delim(x = DAR,delim = "\t",file = "../rds/ATAC_DAR.bed")
```

### Running rGREAt

* Input: DAR region
* Manual input of the DAR.bed file into GREAT v4.0.4 (http://great.stanford.edu/public/html/index.php)
* Output: GO process


## Plotting figure for section one

```{r}

library(ggplot2)
library(ggpubr)

# ---- RNA Volcano Plot ----

Volcano_plot_modification = function(resdata,L2FC,padj_value){
  
  # Adding the conditions for adding of colours
  resdata$con = "non-significant"
  resdata$con[resdata$log2FoldChange > L2FC & resdata$padj < padj_value] = "upregulated"
  resdata$con[resdata$log2FoldChange < -(L2FC) & resdata$padj < padj_value] = "downregulated"
  
  # Deseq2 reports all padj < 1e-303 as 0 so need to be corrected
  resdata$padj = ifelse(resdata$padj == 0,1e-303,resdata$padj)
  
  return(resdata)
}

plot3 = Volcano_plot_modification(resdata = readRDS("../rds/RNA_resdata.rds"),
                                  L2FC = 2,
                                  padj_value = 0.05)

# Plotting
plot3 = ggplot(plot3,aes(x=log2FoldChange,y=-log10(padj),color = con)) + 
  geom_point() +
  theme_classic() +                                                                  
  theme(text = element_text(size = 15),
        legend.position = "none",
        rect = element_rect(fill = "transparent")
        ) +
  scale_colour_manual(values = c("#2191A8","#C0C0C0","#2191A8")) +   # Change the colour of dots
  labs(x = expression("Change in Gene Expression (Log"[2]*" Fold Change)"),
       y = expression("-log"[10]*"(p adjusted)"),
       color = "Gene status") + # naming
  scale_y_continuous(expand = c(0,0),
                     limits = c(0,310))   

# ---- ATAC Volcano Plot ----

Volcano_plot_modification = function(resdata,L2FC,padj_value){
  
  # Adding the conditions for adding of colours
  resdata$con = "non-significant"
  resdata$con[resdata$Fold > L2FC & resdata$FDR < padj_value] = "upregulated"
  resdata$con[resdata$Fold < -(L2FC) & resdata$FDR < padj_value] = "downregulated"
  
  # Deseq2 reports all padj < 1e-303 as 0 so need to be corrected
  resdata$FDR = ifelse(resdata$FDR == 0,1e-303,resdata$FDR)
  
  return(resdata)
}

ATAC_report = Volcano_plot_modification(resdata = readRDS("../rds/ATAC_regdata.rds"),L2FC = 2,padj_value = 0.05)

# Plotting
plot4 = ggplot(ATAC_report,aes(x=Fold,y=-log10(FDR),color = con)) + 
  geom_point() +
  theme_classic() +                                                                  
  theme(text = element_text(size = 15),
        legend.position = "none",
        rect = element_rect(fill = "transparent")) +
  scale_colour_manual(values = c("#E7524C","#C0C0C0","#E7524C")) +   # Change the colour of dots
  labs(x = expression("Change in Chromatin Accessibility (Log"[2]*" Fold Change)"),
       y = expression("-log"[10]*"(p adjusted)"),
       color = "Gene status") + #naming
  scale_y_continuous(expand = c(0,0),limits = c(0,75))  

# ---- RNA Bar plot----

plot2 = readRDS("../Step 3 - Differential Analysis/RNA_ORA.rds")
plot2 = plot2[plot2$ONTOLOGY == "BP",]
plot2 = plot2[1:15,]

plot2 = 
ggplot(data= plot2, aes(x= Count, y = reorder(Description, -p.adjust))) +
  geom_bar(stat="identity",fill = "#2191A8") +
  scale_y_discrete(position = "left") +
  scale_x_reverse() +
  theme(legend.position="none",
        axis.title.y=element_blank(), 
        axis.title.x=element_blank(),
        axis.text.y= element_text(size=12),
        axis.text.x = element_text(size=12),
        axis.line=element_line(color="gray"),
        axis.line.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()
        )


# ---- ATAC barplot ----

plot1 = read.delim("../rds/ATAC_BP.tsv",skip = 3)
plot1 = plot1[1:15,]

plot1 = 
ggplot(data=plot1, aes(x = ObsGenes, y = reorder(Desc, -BinomFdrQ))) +
  geom_bar(stat="identity",fill = "#E7524C") +
  scale_y_discrete(position = "right") +
  theme(legend.position="none",
        axis.title.y=element_blank(), 
        axis.title.x=element_blank(),
        axis.text.y= element_text(size=12),
        axis.text.x = element_text(size=12),
        axis.line=element_line(color="gray"),
        axis.line.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()
  )

# ---- Final ----

Figure1 = ggpubr::ggarrange(ggarrange(plot3,plot4,labels = c("A","B")),
                            ggarrange(plot2,plot1),
                            nrow = 2,labels = c("","C")
                            )

Figure1 = annotate_figure(p = Figure1,bottom = text_grob("Gene count", size = 14))
  
ggsave(plot = Figure1,filename = "../Final/Figure1.png",height = 7,width = 14)

```

# Section Two

## Creation of genomic annotation

### This section is used to create promoter regions with the size of 1kb to -0,5kb of TSS

* Input: gtf file
* Output: promoter rds

```{r}
# Script to create annotations used in this study

library(rtracklayer)
library(plyranges)
library(GenomicFeatures)
library(GenomicRanges)

#### Identify promoters ####

# The definition of promoter is 1kb upstream and .5kb downstream of TSS

promoter_gtf = function(gtf_file) {
  
  # Importing the gtf file
  temp <- rtracklayer::import.gff(gtf_file)
  
  # Removing none related information
  temp = temp[(elementMetadata(temp)[,"type"] == "transcript")]
  temp = temp[(elementMetadata(temp)[,"transcript_type"] != "retained_intron")]
  temp = temp[(elementMetadata(temp)[,"transcript_type"] != "nonsense_mediated_decay")]
  
  # Making the TSS as start and end of each transcript
  start(temp) <- ifelse(strand(temp) == "+",
                        yes = start(temp),
                        no = end(temp))
  end(temp) <- ifelse(strand(temp) == "+",
                      yes = start(temp),
                      no = end(temp))
  
  # Getting the promoter regions
  temp = plyranges::flank_upstream(x = temp,width = 1500)
  temp = plyranges::shift_downstream(x = temp,shift = 500)
  
  return(temp)
}

promoter = promoter_gtf("../../Supporting_file/gencode.v43.primary_assembly.annotation.gtf.gz")
saveRDS(promoter,"../Step 2 - Annotation Curation/promoter.rds")
```

### This section is used to create exon and intron using gtf file

* Input: gtf file
* Output: intron rds, exon rds

```{r}
# Script to create annotations used in this study

library(rtracklayer)
library(plyranges)
library(GenomicFeatures)
library(GenomicRanges)

#### Identify exon and intron ####

gene_gtf = GenomicFeatures::makeTxDbFromGFF("../Supporting file/gencode.v43.primary_assembly.annotation.gtf.gz")

exon <- exonsBy(gene_gtf,by = "tx")
intron <- GenomicFeatures::intronsByTranscript(gene_gtf)

saveRDS(exon,"../Step 2 - Annotation Curation/exon.rds")
saveRDS(intron,"../Step 2 - Annotation Curation/intron.rds")

```

### This section allows the creation of active enhancer for THP-1 cells

* Input: Genhancer bed file; Histone marker
* Output: active enhancer rds

```{r}
# Script to create annotations used in this study

library(rtracklayer)
library(plyranges)
library(GenomicFeatures)
library(GenomicRanges)

# Some genes share multiple gene name so we will need to also use ensembl id as a confirmation
active_enhancer = function(geneHancer,H3K27ac) {
  
  # Importing Enhancer regions
  enhancer = read.delim(geneHancer,header = FALSE,col.names = c("chr","start","end","Genhancer","gene_name","tech"))
  enhancer = GRanges(enhancer)
  
  # Importing H3K27ac markers
  marker = rtracklayer::import.bed(H3K27ac)
  #marker = read.delim(marker,header = FALSE,col.names = c("chr","start","end"))
  
  # Identify active enhancer
  active_enhancer = IRanges::subsetByOverlaps(x = enhancer,ranges = marker)
  
  return(active_enhancer)
}

active_enhancer = active_enhancer(geneHancer = "../Supporting file/geneHancer_format fix.bed",
                                  H3K27ac = "../Step 2 - Annotation Curation/THP1 histone marker/both_marked.bed")

saveRDS(active_enhancer,"../Step 2 - Annotation Curation/THP1_enhancer.rds")
```

## Crosslinking the gene expression changes and chromatin regions

* Input: rds of different chromatin regions
* Output: rds of enhancer and promoter with DAR in it

``` {r}
library(GenomicRanges)
DAR = GRanges(readRDS("../Step 3 - Differential Analysis/ATAC_DAR.rds"))

annotate_region = function(region,event){
  
  region = readRDS(region)
  
  if(class(region) == "CompressedGRangesList") {
    region = as.data.frame(region)
    region = GRanges(region)
  }
  if(class(event) == "data.frame") {
    row.names(event) = 1:nrow(event)
    event = GRanges(event)
  }
  result = annotatr::annotate_regions(regions = region,annotations = event)
  result = as.data.frame(result)
  return(result)
}

enhancer_ATAC = annotate_region("../Step 2 - Annotation Curation/THP1_enhancer.rds",DAR)
promoter_ATAC = annotate_region("../Step 2 - Annotation Curation/promoter.rds",DAR)
intron_ATAC = annotate_region("../Step 2 - Annotation Curation/intron.rds",DAR)
exon_ATAC = annotate_region("../Step 2 - Annotation Curation/exon.rds",DAR)

saveRDS(enhancer_ATAC,"../Step 3 - Differential Analysis/ATAC_DAR_enhancer.rds")
saveRDS(promoter_ATAC,"../Step 3 - Differential Analysis/ATAC_DAR_promoter.rds")
saveRDS(intron_ATAC,"../Step 3 - Differential Analysis/ATAC_DAR_intron.rds")
saveRDS(exon_ATAC,"../Step 3 - Differential Analysis/ATAC_DAR_exon.rds")
```

## Isolating interesting information

```{r}
library(tidyverse)

#### Combine the annotations ####

enhancer_DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR_enhancer.rds")
promoter_DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR_promoter.rds")
DEG = readRDS("../Step 3 - Differential Analysis/RNA_DEGs.rds")

# Combining the information of promoter and enhancer into one
Reg_DAR = rbind(promoter_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name ="promoter"),enhancer_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name = "enhancer"))
Reg_DAR = Reg_DAR %>% distinct(annot.seqnames,gene_name,annot.Fold,name,.keep_all = T)

#### Isolate DEG results with regulators experiencing accessibility changes ####

Reg_DAR_RNA = DEG[DEG$gene_name %in% Reg_DAR$gene_name,]
Reg_DAR_RNA = Reg_DAR[Reg_DAR$gene_name %in% DEG$gene_name,]
saveRDS(Reg_DAR_RNA$gene_name,"../Step 3 - Differential Analysis/DEG&DARinEnhPro.rds")

Reg_RNA_DAR = Reg_DAR[Reg_DAR$gene_name %in% DEG$gene_name,]
#### Over-representation analysis of DEGs with RE ####

ORA_RNA_DEGAcc = ORA_GO(DEG_gene = Reg_DAR_RNA,background = expressed)

#### Isolate region of interest that would be interesting to run footprint ####

extract_gene = function(GO_term,term) {
  temp = GO_term@result
  temp = temp[temp$Description == term,]
  temp = strsplit(as.character(temp$geneID), "/")[[1]]
  return(temp)
}

gene = extract_gene(GO_term = ORA_RNA_DEGAcc,term = "inflammatory response")

enhancer_peak = annotate_region(region = active_enhancer,event = rtracklayer::import.bed("./Genrich/All_peaks_merged.bed"))
enhancer_peak = enhancer_peak[enhancer_peak$gene_name %in% gene,]

promoter_peak = annotate_region(region = promoter,event = rtracklayer::import.bed("./Genrich/All_peaks_merged.bed"))
promoter_peak = promoter_peak[promoter_peak$gene_name %in% gene,]

z = rbind(enhancer_peak %>% dplyr::select(annot.seqnames:annot.end),
          promoter_peak %>% dplyr::select(annot.seqnames:annot.end)) %>% distinct(annot.seqnames,annot.start,annot.end)
rtracklayer::export.bed(z,"Input_TF_Genrich.bed")
```

## Plotting

Plot for barplot for regions
```{r}
library(tidyverse)
library(GenomicRanges)
library(annotatr)

DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR.rds")

annotate_region = function(region,event){
  
  region = readRDS(region)
  
  if(class(region) == "CompressedGRangesList") {
    region = as.data.frame(region)
    region = GRanges(region)
  }
  if(class(event) == "data.frame") {
    row.names(event) = 1:nrow(event)
    event = GRanges(event)
  }
  result = annotatr::annotate_regions(regions = region,annotations = event)
  result = as.data.frame(result)
  return(result)
}

enhancer_ATAC = annotate_region("../Step 2 - Annotation Curation/THP1_enhancer.rds",DAR)
promoter_ATAC = annotate_region("../Step 2 - Annotation Curation/promoter.rds",DAR)
intron_ATAC = annotate_region("../Step 2 - Annotation Curation/intron.rds",DAR)
exon_ATAC = annotate_region("../Step 2 - Annotation Curation/exon.rds",DAR)

DAR$con = "Intergenic"
DAR$con[DAR$Start %in% intron_ATAC$annot.start] = "Intron"
DAR$con[DAR$Start %in% exon_ATAC$annot.start] = "Exon"
DAR$con[DAR$Start %in% enhancer_ATAC$annot.start] = "Enhancer"
DAR$con[DAR$Start %in% promoter_ATAC$annot.start] = "Promoter"

DAR = data.frame(table(DAR$con),con = "1")
DAR$Var1 = factor(DAR$Var1,levels = c("Enhancer","Promoter","Exon","Intron","Intergenic"))

plot1 = ggplot(data=DAR,aes(x = Freq,fill = Var1,y = con)) +
  geom_bar(stat = "identity",position = "stack") +
  scale_y_discrete(position = "right") +
  labs(fill = "Annotation") +
  scale_fill_manual(values = c("forestgreen","#2191A8","goldenrod","#E7524C",'grey')) +
  theme(#legend.position="none",
        axis.title.y=element_blank(), 
        axis.title.x=element_blank(),
        axis.text.y= element_blank(),
        axis.text.x = element_text(size=12),
        axis.line=element_line(color="gray"),
        axis.line.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()
  )

```

Plot for barplot for regions with DEGs
```{r}
library(tidyverse)
library(GenomicRanges)
library(annotatr)

DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR.rds")

annotate_region = function(region,event){
  
  region = readRDS(region)
  
  if(class(region) == "CompressedGRangesList") {
    region = as.data.frame(region)
    region = GRanges(region)
  }
  if(class(event) == "data.frame") {
    row.names(event) = 1:nrow(event)
    event = GRanges(event)
  }
  result = annotatr::annotate_regions(regions = region,annotations = event)
  result = as.data.frame(result)
  return(result)
}

enhancer_ATAC = annotate_region("../Step 2 - Annotation Curation/THP1_enhancer.rds",DAR)
promoter_ATAC = annotate_region("../Step 2 - Annotation Curation/promoter.rds",DAR)
intron_ATAC = annotate_region("../Step 2 - Annotation Curation/intron.rds",DAR)
exon_ATAC = annotate_region("../Step 2 - Annotation Curation/exon.rds",DAR)

DAR$con = "Intergenic"
DAR$con[DAR$Start %in% intron_ATAC$annot.start] = "Intron"
DAR$con[DAR$Start %in% exon_ATAC$annot.start] = "Exon"
DAR$con[DAR$Start %in% enhancer_ATAC$annot.start] = "Enhancer"
DAR$con[DAR$Start %in% promoter_ATAC$annot.start] = "Promoter"

DAR$con

DAR = data.frame(table(DAR$con),con = "1")
DAR$Var1 = factor(DAR$Var1,levels = c("Enhancer","Promoter","Exon","Intron","Intergenic"))

plot1 = ggplot(data=DAR,aes(x = Freq,fill = Var1,y = con)) +
  geom_bar(stat = "identity",position = "stack") +
  scale_y_discrete(position = "right") +
  labs(fill = "Annotation") +
  scale_fill_manual(values = c("forestgreen","#2191A8","goldenrod","#E7524C",'grey')) +
  theme(#legend.position="none",
        axis.title.y=element_blank(), 
        axis.title.x=element_blank(),
        axis.text.y= element_blank(),
        axis.text.x = element_text(size=12),
        axis.line=element_line(color="gray"),
        axis.line.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()
  )

```

Plot for gene number
```{r}
library(tidyverse)
library(GenomicRanges)

annotate_region = function(region,event){
  
  region = readRDS(region)
  
  if(class(region) == "CompressedGRangesList") {
    region = as.data.frame(region)
    region = GRanges(region)
  }
  if(class(event) == "data.frame") {
    row.names(event) = 1:nrow(event)
    event = GRanges(event)
  }
  result = annotatr::annotate_regions(regions = region,annotations = event)
  result = as.data.frame(result)
  return(result)
}

DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR.rds")
DEG = readRDS("../Step 3 - Differential Analysis/RNA_DEGs.rds")
enhancer_ATAC = annotate_region("../Step 2 - Annotation Curation/THP1_enhancer.rds",DAR)
promoter_ATAC = annotate_region("../Step 2 - Annotation Curation/promoter.rds",DAR)

# DAR in both enhancer and promoter would be classified as promoter
enhancer_ATAC = enhancer_ATAC[!paste0(enhancer_ATAC$annot.start,"_",enhancer_ATAC$gene_name) %in% paste0(promoter_ATAC$annot.start,"_",promoter_ATAC$gene_name),]

x = unique(c(enhancer_ATAC$gene_name,promoter_ATAC$gene_name))
x = data.frame(x)

x$con = "Enhancer"
x$con[x$x %in% promoter_ATAC$gene_name] = "Promoter"
x$con[x$x %in% enhancer_ATAC$gene_name & x$x %in% promoter_ATAC$gene_name] = "Both"
x$con[!x$x %in% DEG$gene_name] = "Intron/Intergenic"

#x = data.frame(table(x$con),temp = "1")
#x$Var1 = factor(x$Var1,levels = c("Intron/Intergenic","Enhancer","Both","Promoter"))
x = x[x$con != "Intron/Intergenic",]


plot2 = ggplot(data = x,aes(x = Freq,fill = Var1,y = temp)) +
  geom_bar(stat = "identity",position = "stack") +
  scale_y_discrete(position = "right") +
  labs(fill = "Annotation") +
  scale_fill_manual(values = c('grey','blue','green','red')) +
  theme(#legend.position="none",
        axis.title.y=element_blank(), 
        axis.title.x=element_blank(),
        axis.text.y= element_blank(),
        axis.text.x = element_text(size=12),
        axis.line=element_line(color="gray"),
        axis.line.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.background=element_blank(),
        panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        plot.background=element_blank()
  )
```


# Section Three

## Concentration Preparation

### Running RNA_seq analysis of 3ng/ml

* Input: tx2gene.rds; salmon quant file; gtf_file.rds
* Running the differential analysis using salmon results for low concentration from Bronwyn
* Output: resdata.rds; expressed.rds; DEG.rds

```{r}
# Script to run the differential analysis for RNA-seq

library("DESeq2")
library("tximport")
library("tidyverse")

#### Function One: Running DESeq2  ####

dds_creation = function(files,tx2gene,input_condition){
  
  # Import files into tximport
  txi.salmon <- tximport(files = files,
                         type = "salmon",
                         txOut = FALSE,
                         tx2gene = tx2gene,
                         ignoreAfterBar = TRUE
                         )
  
  # Creating a table with the metainfo
  sampleTable <- data.frame(condition = input_condition)

  # Combining the metainfo with the tximport results
  rownames(sampleTable) <- colnames(txi.salmon$counts) 

  # Importing information into DESeq2 format
  dds <- DESeqDataSetFromTximport(txi = txi.salmon, 
                                  colData = sampleTable,
                                  design = ~ condition)

  # Running the differential expression analysis 
  dds <- DESeq(object = dds,
               test = 'Wald')
  
  return(dds)
  
}

#### Function Two: Extracting usable results and annotate ####

resdata_creation <- function(dds,gtf_file = FALSE,gtf_rds = FALSE){
  
  # Getting the results out of DEseq2
  result = lfcShrink(dds = dds,
                     coef = "condition_treated_vs_control",
                     type = "apeglm"
  )
  
  gtf_file <- readRDS(gtf_rds)
  
  # Putting all important information into a single object
  res <- merge(as.data.frame(result),
               as.data.frame(counts(dds,normalized=TRUE)),
               by = "row.names",
               sort = FALSE)
  
  # Putting the gene symbol with the gene id
  res <- merge(x = res,
               y = gtf_file,
               by.x = "Row.names",
               by.y = "gene_id")
  
  # Getting gene stable version number
  res$Row.names <- gsub("(ENSG[0-9]+)\\.[0-9]+", "\\1",res$Row.names) 
  
  return(res)
}

############### Running differential analysis ##################

dds = dds_creation(files = c("../Step 1 Time/Input/UNT_0_1/quant.sf",
                             "../Step 1 Time/Input/UNT_0_2/quant.sf",
                             "../Step 1 Time/Input/UNT_0_3/quant.sf",
                             "../Step 1 Time/Input/PMA_24_1/quant.sf",
                             "../Step 1 Time/Input/PMA_24_2/quant.sf",                            
                             "../Step 1 Time/Input/PMA_24_3/quant.sf"),
                   tx2gene = tx2gene,
                   input_condition = factor(rep(c("control", "treated"),each = 3)))

resdata_low = resdata_creation(dds = dds,
                              gtf_rds = "../../Paper 1 version One/Step 3 - Differential Analysis/gtf_file.rds")

expressed = resdata[resdata$baseMean > 20,]

DEG = expressed[expressed$padj < 0.05 & abs(expressed$log2FoldChange) > 2,]

saveRDS(resdata,"../Step 3 - Differential Analysis/THP1/low_resdata.rds")
saveRDS(expressed,"../Step 3 - Differential Analysis/THP1/low_expressed.rds")
saveRDS(DEG,"../Step 3 - Differential Analysis/THP1/low_DEGs.rds")
```

### Running RNA_seq analysis of 80ng/ml

* Input: The count file from the study
* 
* Output: resdata.rds; expressed.rds; DEG.rds

```{r}

library(tidyverse)
library(DESeq2)

count = read.delim("../Step 4 - Concentration/80ng/gene_count.xls")
count = column_to_rownames(count,var = "gene_id")
dds <- DESeqDataSetFromMatrix(countData = count,
                              colData = data.frame(condition = c(rep(c("control", "treated"),each = 3))),
                              design = ~ condition)

# Running the differential expression analysis 
dds <- DESeq(object = dds,
             test = 'Wald')

# Getting the results out of DEseq2
result = lfcShrink(dds = dds,
                   coef = "condition_treated_vs_control",
                   type = "apeglm"
)


res <- merge(as.data.frame(result),
             as.data.frame(counts(dds,normalized=TRUE)),
             by = "row.names",
             sort = FALSE)

library('biomaRt')
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol"),values=res$Row.names,mart= mart)
resdata_moderate <- merge(x = res,
                 y = G_list,
                 by.x = "Row.names",
                 by.y = "ensembl_gene_id",all=T)
x = openxlsx::read.xlsx("../Step 4 - Concentration/80ng/Allgene_info.xls.xlsx")
resdata_moderate = merge(x = resdata_moderate,
                         y = x[,1:2],
                         by.x = "Row.names",
                         by.y = "gene_id")

saveRDS(resdata_moderate,file = "../Step 4 - Concentration/resdata_80ng.rds")

```

## Cell line Preparation

### Running RNA-seq analysis of HL60 cell line

## Elimating genes based on concentration

* Input:low concentration; moderate concentration; high concentration
* Output: Concentration independents DEGs with DAR activities

```{r}
# Importing and preparing the results

low = readRDS("../Step 4 - Concentration/resdata_5nm.rds")
moderate = readRDS("../Step 4 - Concentration/resdata_80ng.rds")
high = readRDS("../Step 3 - Differential Analysis/RNA_resdata.rds")

# Only genes within the topn 16000 expression are considered to be expressed to account for the sample variation
high = high %>% top_n(16000,baseMean)
moderate = moderate %>% top_n(16000,baseMean)
low = low %>% top_n(16000,baseMean)

# Genes of interest from step 3
geneOfInterest = readRDS("../Step 3 - Differential Analysis/DEG&DARinEnhPro.rds")
geneOfInterest = geneOfInterest[geneOfInterest %in% low$gene_name & geneOfInterest %in% moderate$gene_name]

# Combining the results into one dataframe
Combine = merge(low,high,by = c("gene_name","Row.names"),all=TRUE,suffixes = c("low","high"))
Combine = merge(Combine,moderate,by = c("gene_name"))
Combine = Combine[Combine$gene_name %in% geneOfInterest,]
Combine = Combine %>% dplyr::select(gene_name,log2FoldChangelow,log2FoldChange,log2FoldChangehigh)
colnames(Combine) = c("gene","low","moderate","high")
saveRDS(Combine,"../Final Object/Concentration_DAR.rds")
```

## Elimanating based on cell line

* Input:Concentration independents DEGs with DAR activities; HL60 gene expression
* Output: Concentration independents DEGs and cell type independent with DAR activities

```{r}
HL60_RNA = readRDS("../Step 5 - cell line/HL60/RNA_resdata.rds")
HL60_RNA = HL60_RNA %>% top_n(16000,baseMean)

# Input of gene of interest
geneOfInterest = readRDS("../Final Object/Concentration_DAR.rds")
geneOfInterest = merge(geneOfInterest,HL60_RNA,by.x = "gene",by.y = "gene_name")
geneOfInterest = geneOfInterest %>% dplyr::select(gene:high,log2FoldChange)
colnames(geneOfInterest)[colnames(geneOfInterest) == "log2FoldChange"] = "HL-60"

saveRDS(geneOfInterest,"../Final Object/Concentration_CellLine_DAR.rds")

geneOfInterest = geneOfInterest[rowSums(geneOfInterest > 2 | geneOfInterest < -2) == 5,]


saveRDS(geneOfInterest,"../Final Object/ConcentrationDEG_CellLineDEG_DAR.rds")
```

## Plotting

```{r}
# Dividing the gene expression pattern into groups
Combine$Group = "E"
Combine$Group[Combine$low < 0 & Combine$moderate < 0 & Combine$high < 0] = "D"
Combine$Group[Combine$low > 0 & Combine$moderate > 0 & Combine$high > 0] = "C"
Combine$Group[Combine$low < -2 & Combine$moderate < -2 & Combine$high < -2] = "B"
Combine$Group[Combine$low > 2 & Combine$moderate > 2 & Combine$high > 2] = "A"

saveRDS(Combine,"../Step 4 - Concentration/Concentration&Accessibility.rds")

# Plotting the result
library(ComplexHeatmap)
library(circlize)

PlotFrame = Combine
rownames(PlotFrame) = PlotFrame$gene
split = factor(toupper(PlotFrame$Group), levels = toupper(letters[1:5]))

PlotFrame = data.matrix(PlotFrame)
plot1 = Heatmap(PlotFrame[,2:4],
        name = "Log2FoldChange",heatmap_legend_param = list(legend_direction = "vertical"),
        row_split = split,cluster_rows = FALSE,
        col = colorRamp2(c(-2,0,2), c( "#B9181A","white","#4995C6")),
        show_row_dend = FALSE,show_column_dend = FALSE,show_row_names = FALSE)
```

```{r}
library(tidyverse)
library(clusterProfiler)
library(org.Hs.eg.db)
# Input of gene of interest
geneOfInterest = readRDS("../Step 4 - Concentration/Concentration&Accessibility.rds")
geneOfInterest = merge(geneOfInterest,HL60_RNA,by.x = "gene",by.y = "gene_name")
geneOfInterest = geneOfInterest[geneOfInterest$Group %in% c("A","B"),]
geneOfInterest = geneOfInterest %>% dplyr::select(gene,log2FoldChange,high,moderate,low)
colnames(geneOfInterest)[colnames(geneOfInterest) == "log2FoldChange"] = "HL-60"

# Input of gene expression firnoamtion
Combine = geneOfInterest
Combine$Group = "C"
Combine$Group[Combine$low < -2 & Combine$moderate < -2 & Combine$high < -2 & Combine$`HL-60` < -2] = "B"
Combine$Group[Combine$low > 2 & Combine$moderate > 2 & Combine$high > 2 & Combine$`HL-60` > 2] = "A"

PlotFrame = Combine
rownames(PlotFrame) = PlotFrame$gene
split = factor(toupper(PlotFrame$Group), levels = toupper(letters[1:3]))

saveRDS(Combine,"./Combine.rds")

PlotFrame = data.matrix(PlotFrame)
plot2 = Heatmap(PlotFrame[,2:5],
        name = "log2FoldChange",
        row_split = split,cluster_rows = FALSE,row_gap = unit(2, "mm"),
        column_split = 2,column_gap = unit(5, "mm"),
        col = colorRamp2(c(-2,0,2), c( "#B9181A","white","#4995C6")),
        show_row_dend = FALSE,show_column_dend = FALSE,show_row_names = FALSE)
plot2 = Heatmap(PlotFrame[,2:5],
        name = "log2FoldChange",
        row_split = split,cluster_rows = FALSE,row_gap = unit(2, "mm"),
        column_split = 2,column_gap = unit(5, "mm"),
        col = colorRamp2(c(-2,0,2), c( "#B9181A","white","#4995C6")),
        show_row_dend = FALSE,show_column_dend = FALSE,show_row_names = FALSE)






x = Combine[Combine$Group %in% c("A","B"),]
x = clusterProfiler::enrichGO(x$gene,OrgDb = org.Hs.eg.db,keyType = "SYMBOL",ont = "BP")
plot3 = barplot(x)

ggarrange(plot1,plot2,plot3,ncol = 1)
```

# Section Four

## Running ATAC-seq analysis of HL60 cell line

* Input: Genrich peak; BAM file
* Output: regdata.rds; DAR.rds; DAR.bed

```{r}
# Script to run the differential analysis for ATAC-seq

########### Running preparation for ATAC-seq differential analysis #############
################################################################################

library(DiffBind)

# Putting in the metadata
sampleTbl = data.frame(sampleID = c('control1','control2','treated1','treated2'),
                       Cell = 'HL60',
                       Factor = 'PMA',
                       Condition= c('Control','Control','Treated','Treated'),
                       Replicate = c("1","2","1","2"),
                       bamReads = c('../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213963.dedup.quality.bam',
                                    '../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213964.dedup.quality.bam',
                                    #'../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213965.dedup.quality.bam',
                                    '../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213975.dedup.quality.bam',
                                    '../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213976.dedup.quality.bam'
                                    #'../Step 1 - Preprocessing/HL60_bowtie2/4. Final/SRR3213977.dedup.quality.bam'
                                    ),
                       Peaks = c('../Step 1 - Preprocessing/HL60_Genrich/SRR3213963.narrowPeak',
                                 '../Step 1 - Preprocessing/HL60_Genrich/SRR3213964.narrowPeak',
                                 #'../Step 1 - Preprocessing/HL60_Genrich/SRR3213965.narrowPeak',
                                 '../Step 1 - Preprocessing/HL60_Genrich/SRR3213975.narrowPeak',
                                 '../Step 1 - Preprocessing/HL60_Genrich/SRR3213976.narrowPeak'
                                 #'../Step 1 - Preprocessing/HL60_Genrich/SRR3213977.narrowPeak'
                                 ),
                       ScoreCol = 5,
                       LowerBetter = FALSE
)

# Importing the data
ATAC <- dba(sampleSheet=sampleTbl)

# Perform count on each peak using the BAM file
ATAC_count <- dba.count(ATAC,minOverlap = 2,
                        score = DBA_SCORE_NORMALIZED,
                        bUseSummarizeOverlaps = TRUE
)

saveRDS(ATAC_count,"../Step 5 - cell line/ATAC_count.rds")
remove(sampleTbl,ATAC)


################################ Running DiffBind ##############################
################################################################################

# Differential accessibility analysis
ATAC_contrast <- dba.contrast(ATAC_count, categories=DBA_CONDITION,minMembers = 2)

# Perform differential affinitiy analysis
ATAC_analyze <- dba.analyze(ATAC_contrast,method = DBA_ALL_METHODS)

# Extracting the results
ATAC_report <- dba.report(DBA = ATAC_analyze,
                          DataType = DBA_DATA_FRAME,
                          method = DBA_DESEQ2,
                          contrast = 1,
                          th = 1)

saveRDS(ATAC_report,"../Step 5 - cell line/ATAC_regdata.rds")
remove(ATAC_count,ATAC_contrast,ATAC_analyze)

# Extracting usable regions
DAR = ATAC_report %>% dplyr::filter(abs(Fold) > 2 & FDR < 0.01)
row.names(DAR) = 1:nrow(DAR)
saveRDS(DAR,"../Step 5 - cell line/ATAC_DAR.rds")

######################### Accessibility of regions #############################
################################################################################

DAR = GRanges(DAR)

annotate_region = function(region,event){
  
  region = readRDS(region)
  
  if(class(region) == "CompressedGRangesList") {
    region = as.data.frame(region)
    region = GRanges(region)
  }
  if(class(event) == "data.frame") {
    row.names(event) = 1:nrow(event)
    event = GRanges(event)
  }
  result = annotatr::annotate_regions(regions = region,annotations = event)
  result = as.data.frame(result)
  return(result)
}

enhancer_ATAC = annotate_region("../Step 2 - Annotation Curation/THP1_enhancer.rds",DAR)
promoter_ATAC = annotate_region("../Step 2 - Annotation Curation/promoter.rds",DAR)
intron_ATAC = annotate_region("../Step 2 - Annotation Curation/intron.rds",DAR)
exon_ATAC = annotate_region("../Step 2 - Annotation Curation/exon.rds",DAR)

saveRDS(enhancer_ATAC,"../Step 5 - cell line/ATAC_DAR_enhancer.rds")
saveRDS(promoter_ATAC,"../Step 5 - cell line/ATAC_DAR_promoter.rds")

```

### Locaing ones just in THP-1 cells

```{r}

geneOfInterest = readRDS("../Final Object/ConcentrationDEG_CellLineDEG_DAR.rds")
enhancer_DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR_enhancer.rds")
promoter_DAR = readRDS("../Step 3 - Differential Analysis/ATAC_DAR_promoter.rds")

# Combining the information of promoter and enhancer into one
Reg_DAR = rbind(promoter_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name ="promoter"),enhancer_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name = "enhancer"))
Reg_DAR = Reg_DAR %>% distinct(annot.seqnames,gene_name,annot.Fold,name,.keep_all = T)

# Looking at specfici genes
Reg_DAR = Reg_DAR %>% filter(gene_name %in% geneOfInterest$gene)

saveRDS(Reg_DAR,"../Final Object/ConcentrationDEG_CellLineDEG_THP1DAR.rds")
```

### locating ones just in HL60 cells

```{r}

geneOfInterest = readRDS("../Final Object/ConcentrationDEG_CellLineDEG_DAR.rds")
enhancer_DAR = readRDS("../Step 5 - cell line/ATAC_DAR_enhancer.rds")
promoter_DAR = readRDS("../Step 5 - cell line/ATAC_DAR_promoter.rds")

# Combining the information of promoter and enhancer into one
Reg_DAR = rbind(promoter_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name ="promoter"),enhancer_DAR %>% dplyr::select(annot.seqnames:annot.end,gene_name,annot.Fold) %>% mutate(name = "enhancer"))
Reg_DAR = Reg_DAR %>% distinct(annot.seqnames,gene_name,annot.Fold,name,.keep_all = T)

# Looking at specfici genes
Reg_DAR = Reg_DAR %>% filter(gene_name %in% geneOfInterest$gene)

saveRDS(Reg_DAR,"../Final Object/ConcentrationDEG_CellLineDEG_HL60DAR.rds")
```

### In both

```{r}
THP1 = readRDS("../Final Object/ConcentrationDEG_CellLineDEG_THP1DAR.rds")
HL60 = readRDS("../Final Object/ConcentrationDEG_CellLineDEG_HL60DAR.rds")

Both = THP1[THP1$gene_name %in% HL60$gene_name,]

library(plyranges)
library(tidyverse)

Both_exact = plyranges::find_overlaps(GRanges(THP1),GRanges(HL60))
Both_exact = data.frame(Both_exact)
Both_exact = Both_exact %>% distinct(start,end,gene_name.x,.keep_all = T)

```

## Plotting

```{r}
library(ComplexHeatmap)
library(circlize)

bed = Both_exact %>% select(seqnames:end,annot.Fold.x,gene_name.x)
circos.clear()
#circos.initializeWithIdeogram(plotType = c("labels", "axis"))

circos.initializeWithIdeogram()

circos.genomicTrack(bed, panel.fun = function(region, value, ...) {
    circos.genomicPoints(region, value, pch = 16, cex = 0.5, ...)
    i = getI(...)
    circos.lines(CELL_META$cell.xlim, c(i, i), lty = 2, col = "#00000040")
})

#circos.genomicIdeogram()
circos.genomicLabels(bed, labels.column = 5, side = "inside",
    col = as.numeric(factor(bed[[1]])), line_col = as.numeric(factor(bed[[1]])))

```

```{r}
library(ComplexHeatmap)
library(circlize)
library(tidyverse)
gtf = rtracklayer::import.gff("../../Supporting_file/gencode.v43.primary_assembly.annotation.gtf")
y = as.data.frame(gtf)
y = y %>% dplyr::filter(type == "exon")
y = y %>% dplyr::filter(gene_name %in% c("CSF1R","BCL6","CSF1"))
y = y %>% select(gene_name,start,end,transcript_id,exon_number)
colnames(y)=c('gene','start','end','transcript','exon')

x = Both_exact %>% dplyr::filter(gene_name.x %in% c("CSF1R","BCL6","CSF1"))
x = x %>% select(gene_name.x,start,end,annot.Fold.x)
x$exon = 1
colnames(x)=c('gene','start','end','transcript','exon')

y = rbind(y,x)

df = data.frame(
     name  = c("CSF1R","BCL6","CSF1"),
     start = c(150050000, 187721000, 109793000),
     end   = c(150114000, 187987000, 109931000))

circos.clear()

circos.genomicInitialize(y)

circos.track(ylim = c(0, 1), 
    bg.col = c("#FF000040", "#00FF0040", "#0000FF40"), 
    bg.border = NA, track.height = 0.05)

n = max(tapply(y$transcript, y$gene, function(x) length(unique(x))))
circos.genomicTrack(y, ylim = c(0.5, n + 0.5), 
    panel.fun = function(region, value, ...) {
        all_tx = unique(value$transcript)
        for(i in seq_along(all_tx)) {
            l = value$transcript == all_tx[i]
            # for each transcript
            current_tx_start = min(region[l, 1])
            current_tx_end = max(region[l, 2])
            circos.lines(c(current_tx_start, current_tx_end), 
                c(n - i + 1, n - i + 1), col = "#CCCCCC")
            circos.genomicRect(region[l, , drop = FALSE], ytop = n - i + 1 + 0.4, 
                ybottom = n - i + 1 - 0.4, col = "orange", border = NA)
        }
}, bg.border = NA, track.height = 0.4)


bed = Both_exact %>% select(seqnames:end,annot.Fold.x,gene_name.x)
bed = bed %>% dplyr::filter(gene_name.x %in% c("CSF1R","BCL6","CSF1"))

circos.genomicTrack(bed, 
    panel.fun = function(region, value, ...) {
        circos.genomicRect(region, value, ytop.column = 1, ybottom = 0, 
            col = ifelse(value[[1]] > 0, "red", "green"), ...)
        circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = "#00000040")
})

circos.clear()
#circos.initializeWithIdeogram(plotType = c("labels", "axis"))

circos.initializeWithIdeogram()

circos.genomicTrack(bed, panel.fun = function(region, value, ...) {
    circos.genomicPoints(region, value, pch = 16, cex = 0.5, ...)
    i = getI(...)
    circos.lines(CELL_META$cell.xlim, c(i, i), lty = 2, col = "#00000040")
})

#circos.genomicIdeogram()
circos.genomicLabels(bed, labels.column = 5, side = "inside",
    col = as.numeric(factor(bed[[1]])), line_col = as.numeric(factor(bed[[1]])))

```

# Section Five

```{r}
library(plyranges)
library(tidyverse)

x = readRDS("../Oth.Bld.05.AllAg.HL-60.rds")

x = readRDS("../Oth.Bld.05.AllAg.THP-1.rds")

y = rtracklayer::import.bed("../R/test33fsfds3.bed")
y$name = 1:y@elementMetadata@nrows

x = plyranges::find_overlaps(x,y,suffix = c(".ChIP","./Region"))
x = as.data.frame(x)
x$name = str_split_i(x$name.ChIP,";",2)
x$name = str_split_i(x$name,"=",2)
x$name = str_split_i(x$name,"%",1)

x = x %>% distinct(name..Region,name,.keep_all = T)
table(x$name)

```


# Reference


